include/clang-c/CXCppInterOp.h

```diff
@@ -0,0 +1,366 @@
+// NOLINTBEGIN()
+#ifndef LLVM_CLANG_C_CXCPPINTEROP_H
+#define LLVM_CLANG_C_CXCPPINTEROP_H
+
+#include "clang-c/CXErrorCode.h"
+#include "clang-c/CXString.h"
+#include "clang-c/ExternC.h"
+#include "clang-c/Index.h"
+#include "clang-c/Platform.h"
+
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+
+LLVM_CLANG_C_EXTERN_C_BEGIN
+
+/**
+ * \defgroup CPPINTEROP_INTERPRETER_MANIP Interpreter manipulations
+ *
+ * @{
+ */
+
+/**
+ * An opaque pointer representing an interpreter context.
+ */
+typedef struct CXInterpreterImpl* CXInterpreter;
+
+/**
+ * Create a Clang interpreter instance from the given arguments.
+ *
+ * \param argv The arguments that would be passed to the interpreter.
+ *
+ * \param argc The number of arguments in \c argv.
+ *
+ * \returns a \c CXInterpreter.
+ */
+CXInterpreter clang_createInterpreter(const char* const* argv, int argc);
+
+typedef void* TInterp_t;
+
+/**
+ * Bridge between C API and C++ API.
+ *
+ * \returns a \c CXInterpreter.
+ */
+CXInterpreter clang_createInterpreterFromRawPtr(TInterp_t I);
+
+/**
+ * Returns a pointer to the underlying interpreter.
+ */
+void* clang_interpreter_getClangInterpreter(CXInterpreter I);
+
+/**
+ * Returns a \c TInterp_t and takes the ownership.
+ */
+TInterp_t clang_interpreter_takeInterpreterAsPtr(CXInterpreter I);
+
+/**
+ * Undo N previous incremental inputs.
+ */
+enum CXErrorCode clang_interpreter_Undo(CXInterpreter I, unsigned int N);
+
+/**
+ * Dispose of the given interpreter context.
+ */
+void clang_interpreter_dispose(CXInterpreter I);
+
+/**
+ * Describes the return result of the different routines that do the incremental
+ * compilation.
+ */
+typedef enum {
+  /**
+   * The compilation was successful.
+   */
+  CXInterpreter_Success = 0,
+  /**
+   * The compilation failed.
+   */
+  CXInterpreter_Failure = 1,
+  /**
+   * More more input is expected.
+   */
+  CXInterpreter_MoreInputExpected = 2,
+} CXInterpreter_CompilationResult;
+
+/**
+ * Add a search path to the interpreter.
+ *
+ * \param I The interpreter.
+ *
+ * \param dir The directory to add.
+ *
+ * \param isUser Whether the directory is a user directory.
+ *
+ * \param prepend Whether to prepend the directory to the search path.
+ */
+void clang_interpreter_addSearchPath(CXInterpreter I, const char* dir,
```

Comment:
Do we want to shorten these interfaces. Maybe `Cpp_AddSearchPath`?

Reply:
Since we only expose opaque pointers in the C API, I believe libclang uses this naming style to warn users about the type of the input argument. However, for historical reasons, libclang appears to lack strict rules on capitalization. 

The naming convention is like `clang_` + `ClassName` + `methodName` or `clang_` + `functionName`.

Here are some examples:

```
typedef void *CXIndexAction;
CINDEX_LINKAGE CXIndexAction clang_IndexAction_create(CXIndex CIdx);
CINDEX_LINKAGE void clang_IndexAction_dispose(CXIndexAction);

CINDEX_LINKAGE CXIdxClientContainer
clang_index_getClientContainer(const CXIdxContainerInfo *);

CINDEX_LINKAGE unsigned clang_CXXMethod_isDeleted(CXCursor C);

CINDEX_LINKAGE CXString clang_Module_getName(CXModule Module);

CINDEX_LINKAGE CXCursor clang_getCanonicalCursor(CXCursor);

CINDEX_LINKAGE unsigned
clang_PrintingPolicy_getProperty(CXPrintingPolicy Policy, enum CXPrintingPolicyProperty Property);

CINDEX_LINKAGE void
clang_PrintingPolicy_setProperty(CXPrintingPolicy Policy,
                                 enum CXPrintingPolicyProperty Property,
                                 unsigned Value);

CINDEX_LINKAGE CXPrintingPolicy clang_getCursorPrintingPolicy(CXCursor);

CINDEX_LINKAGE void clang_PrintingPolicy_dispose(CXPrintingPolicy Policy);

CINDEX_LINKAGE CXString clang_getCursorPrettyPrinted(CXCursor Cursor, CXPrintingPolicy Policy);
```